compiler/rustc/src/main.rs

```diff
@@ -1,62 +1,57 @@
-//
-// A consequence of not using `GlobalAlloc` (and the `tikv-jemallocator` crate
-// provides an impl of that trait, which is called `Jemalloc`) is that we
-// cannot use the sized deallocation APIs (`sdallocx`) that jemalloc provides.
-// It's unclear how much performance is lost because of this.
+// mimalloc.
 //
 // As for the symbol overrides in `main` below: we're pulling in a static copy
-// of jemalloc. We need to actually reference its symbols for it to get linked.
+// of mimalloc. We need to actually reference its symbols for it to get linked.
 // The two crates we link to here, `std` and `rustc_driver`, are both dynamic
-// libraries. So we must reference jemalloc symbols one way or another, because
+// libraries. So we must reference mimalloc symbols one way or another, because
 // this file is the only object code in the rustc executable.
-#[cfg(feature = "tikv-jemalloc-sys")]
-use tikv_jemalloc_sys as jemalloc_sys;
 
 fn main() {
     // See the comment at the top of this file for an explanation of this.
-    #[cfg(feature = "tikv-jemalloc-sys")]
+    #[cfg(feature = "libmimalloc-sys")]
     {
         use std::os::raw::{c_int, c_void};
 
         #[used]
-        static _F1: unsafe extern "C" fn(usize, usize) -> *mut c_void = jemalloc_sys::calloc;
+        static _F1: unsafe extern "C" fn(usize, usize) -> *mut c_void = libmimalloc_sys::mi_calloc;
         #[used]
         static _F2: unsafe extern "C" fn(*mut *mut c_void, usize, usize) -> c_int =
-            jemalloc_sys::posix_memalign;
+            libmimalloc_sys::mi_posix_memalign;
         #[used]
-        static _F3: unsafe extern "C" fn(usize, usize) -> *mut c_void = jemalloc_sys::aligned_alloc;
+        static _F3: unsafe extern "C" fn(usize, usize) -> *mut c_void =
+            libmimalloc_sys::mi_aligned_alloc;
         #[used]
-        static _F4: unsafe extern "C" fn(usize) -> *mut c_void = jemalloc_sys::malloc;
+        static _F4: unsafe extern "C" fn(usize) -> *mut c_void = libmimalloc_sys::mi_malloc;
         #[used]
-        static _F5: unsafe extern "C" fn(*mut c_void, usize) -> *mut c_void = jemalloc_sys::realloc;
+        static _F5: unsafe extern "C" fn(*mut c_void, usize) -> *mut c_void =
+            libmimalloc_sys::mi_realloc;
         #[used]
-        static _F6: unsafe extern "C" fn(*mut c_void) = jemalloc_sys::free;
+        static _F6: unsafe extern "C" fn(*mut c_void) = libmimalloc_sys::mi_free;
 
-        // On OSX, jemalloc doesn't directly override malloc/free, but instead
+        // On OSX, mimalloc doesn't directly override malloc/free, but instead
         // registers itself with the allocator's zone APIs in a ctor. However,
         // the linker doesn't seem to consider ctors as "used" when statically
         // linking, so we need to explicitly depend on the function.
         #[cfg(target_os = "macos")]
         {
             extern "C" {
-                fn _rjem_je_zone_register();
+                fn _mi_macos_override_malloc();
```

Comment:
Hmm, I did not test this in macOS because I don't have a machine accessible. Windows is not supported because static linking cannot be used with override. Not sure what the next steps are...

Reply:
The results look really good to me, so I think we should try to find a way to fix this, if possible. I am happy to test this, but I am not proficient with macOS either. Maybe @sfackler or @ehuss have ideas what to try next (as you fixed `jemalloc` for macOS earlier)?

Reply:
Also, @thomcc may be interested to give new ideas.

Reply:
You have to enable stats when building or they aren't tracked, (see: https://github.com/microsoft/mimalloc/blob/master/include/mimalloc-types.h#L32-L33). I don't remember if theres support in the mimalloc (or -sys) crates directly, but setting `env CFLAGS="-DMI_STAT=1"` or `env CFLAGS="-DMI_STAT=2"` when building should make `cc` do it, hopefully.

Reply:
Thanks! So `_mi_macos_override_malloc()` should do it but stats may not be enabled correctly in my setup. That would be excellent, I'll try some more runs and report back.

Reply:
@thomcc Thanks, I checked this in detail. In my debug build I enable `MI_DEBUG_FULL` in `CMakeLists.txt` and this will enable `MI_DEBUG=3`definition. This then again will set `MI_STAT=2` [https://github.com/microsoft/mimalloc/blob/master/include/mimalloc-types.h#L421-L427](https://github.com/microsoft/mimalloc/blob/master/include/mimalloc-types.h#L421-L427). I cannot easily see what could go wrong, but to be sure, I'll try with a separate forced setting to `MI_STAT`.

If the statistics are valid, then `mi_macos_override_malloc` seems not to be called, perhaps due to similar reasons `jemalloc` originally did not work, where the linker is optimizing away some needed call paths. We would need to figure out what is still missing for `mimalloc`. My skills need sharpening here, so any hints on how to do it best are welcome :)

Reply:
Do you have your integration code published anywhere? If not I can take a stab, but would like to avoid redundant work.

I think the mi_macos_override_malloc stuff is requires a little bit of extra work. most of the documentatoin mimalloc has about what's required to override on macos is a bit dated (or wrong) too (interposing happens by default, flat namespace should not be required).

Also, the zone override is not intended to be used on its own, but that's fixable.

Reply:
@thomcc Thank you for taking a look! My current changes are just testing and stats debug related, which we discussed above. I do not have anything else outside this PR.

Reply:
Taking more of a look, it's very unsurprising that you can't call _mi_macos_override_malloc directly, it's a static private function. It's also only included when mimalloc is built as a dynamic library (there are various reasons for this, but I think the main one is that mimalloc attempts interposing, and prior to macOS 10.12, this could only be done as a dynamic lib with some DYLD_ environment vars defined).

There are at least a few issues with the way rustc will need to use it. Nothing that seems enormous, thankfully, I'll see if I can come up with a fixed version we can test out, and upstream if it works fine.

The biggest issue is probably that I don't have an easy way to test on older versions of macOS, so will be taking a stab in them. I also don't know how far back rustc supports for the host (I know libstd supports back to 10.7, though). Also, does rustc fork itself during execution? If so, that may complicate things a little, at least in terms of testing.

Reply:
> Also, does rustc fork itself during execution?

No. It only spawns a new thread to run the compilation on.